### PR TITLE
feat: add adaptor to convert net/http handler to hertz request handler

### PR DESCRIPTION
**Pull request by Skyenought**
#### What type of PR is this?

feat
#### Check the PR title.

- [x] This PR title match the format: \<type\>(optional scope): \<description\>
- [x] The description of this PR title is user-oriented and clear enough for others to understand.

#### (Optional) More detail description for this PR(en: English/zh: Chinese).

zh(optional): 添加将 net/http handler 转为 hertz app.HandleFunc 的适配器

### CODE REVIEW

**Review by wangbei98 — APPROVED** [review]

**Review by wangbei98 — APPROVED** [review]

**Review by wangbei98 — APPROVED** [review]

File: pkg/common/adaptor/response.go [hunk 1]
```diff
@@ -88,3 +92,68 @@ func GetCompatResponseWriter(resp *protocol.Response) http.ResponseWriter {
...
+//     so the returned handler will be always slower than manually written
+//     hertz handler.
+//
+// So it is advisable using this function only for quick net/http -> hertz switching.
+// Then manually convert net/http handlers to hertz handlers
+func NewHertzHTTPHandler(h http.Handler) app.HandlerFunc {
+	return func(ctx context.Context, c *app.RequestContext) {
+		req, err := GetCompatRequest(c.GetRequest())
+		if err != nil {
+			hlog.CtxErrorf(ctx, "HERTZ: Get request error: %v", err)
```
Comment (li-jin-gou) [inline pkg/common/adaptor/response.go:130]: need return because throw err

File: pkg/common/adaptor/response.go [hunk 2]
```diff
@@ -88,3 +92,68 @@ func GetCompatResponseWriter(resp *protocol.Response) http.ResponseWriter {
...
+// it has the following drawbacks comparing to using manually written hertz
+// request handler:
+//
+//   - net/http -> hertz handler conversion has some overhead,
+//     so the returned handler will be always slower than manually written
+//     hertz handler.
+//
+// So it is advisable using this function only for quick net/http -> hertz switching.
+// Then manually convert net/http handlers to hertz handlers
+func NewHertzHTTPHandlerFunc(h http.HandlerFunc) app.HandlerFunc {
```
Comment (li-jin-gou) [inline pkg/common/adaptor/response.go:109]: It seems inappropriate to put it in `response.go`, create hanlder.go and handler_test.go

File: pkg/common/adaptor/handler.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+// it has the following drawbacks comparing to using manually written hertz
+// request handler:
+//
+//   - net/http -> hertz handler conversion has some overhead,
+//     so the returned handler will be always slower than manually written
+//     hertz handler.
+//
+// So it is advisable using this function only for net/http -> hertz switching.
+// Then manually convert net/http handlers to hertz handlers
+func NewHertzHTTPHandlerFunc(h http.HandlerFunc) app.HandlerFunc {
```
Comment (li-jin-gou) [inline pkg/common/adaptor/handler.go:43]: Why would you pack an extra layer

Comment (wangbei98) [reply pkg/common/adaptor/handler.go:43]: https://github.com/cloudwego/hertz/pull/253#discussion_r973595291
without this extra layer, caller should do like this:

    index := http.HandlerFunc(pprof.Index)  // convert
    prefixRouter.GET("/", NewHertzHTTPHandler(index))

this layer can help user to convert f(w http.ResponseWriter, r *http.Request) to  http.HandlerFunc

Comment (li-jin-gou) [reply pkg/common/adaptor/handler.go:43]: thanks

File: pkg/common/adaptor/handler_test.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+		for k, expectedV := range expectedHeader {
+			v := r.Header.Get(k)
+			if v != expectedV {
+				t.Fatalf("unexpected header value %q for key %q. Expecting %q", v, k, expectedV)
+			}
+		}
+		w.Header().Set("Header1", "value1")
+		w.Header().Set("Header2", "value2")
+		w.WriteHeader(http.StatusBadRequest)
+		w.Write([]byte("<!doctype html><html>")) //nolint:errcheck
```
Comment (li-jin-gou) [inline pkg/common/adaptor/handler_test.go:87]: 这里可以直接用body

File: pkg/common/adaptor/handler.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+ */
+
+package adaptor
+
+import (
+	"bytes"
+	"context"
+	"net/http"
+
+	"github.com/cloudwego/hertz/internal/bytesconv"
```
Comment (li-jin-gou) [inline pkg/common/adaptor/handler.go:24]: import

**Review by li-jin-gou — DISMISSED** [review]

File: pkg/common/adaptor/handler.go [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+		c.ForEachKey(func(k string, v interface{}) {
+			ctx = context.WithValue(ctx, k, v)
+		})
+
+		h.ServeHTTP(rw, req.WithContext(ctx))
+
+		c.SetStatusCode(rw.h.StatusCode())
+		haveContentType := false
+		for k, vv := range rw.header {
+			if k == "Content-Type" {
```
Comment (FGYFFFF) [inline pkg/common/adaptor/handler.go:79]: "Content-Type" 有个常量const，代码里可以直接用

File: pkg/common/adaptor/handler.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+			// From net/http.ResponseWriter.Write:
+			// If the Header does not contain a Content-Type line, Write adds a Content-Type set
+			// to the result of passing the initial 512 bytes of written data to DetectContentType.
+			l := 512
+			if len(body) < 512 {
+				l = len(body)
+			}
+			c.Response.Header.Set("Content-Type", http.DetectContentType(body[:l]))
+		}
+		c.Response.SetBodyStream(bytes.NewBuffer(body), len(body))
```
Comment (FGYFFFF) [inline pkg/common/adaptor/handler.go:97]: 如果net/http 的handler返回的是"chunk"编码，这块还需要处理吗？还是不考虑这种情况？

Comment (li-jin-gou) [reply pkg/common/adaptor/handler.go:97]: 暂时不考虑

File: pkg/common/adaptor/handler.go [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+)
+
+// NewHertzHTTPHandlerFunc wraps net/http handler to hertz app.HandlerFunc,
+// so it can be passed to hertz server
+//
+// While this function may be used for easy switching from net/http to hertz,
+// it has the following drawbacks comparing to using manually written hertz
+// request handler:
+//
+//   - net/http -> hertz handler conversion has some overhead,
```
Comment (li-jin-gou) [inline pkg/common/adaptor/handler.go:36]: //   - A lot of useful functionality provided by hertz is missing
//     from net/http handler.

补充一下这个吧

File: pkg/common/adaptor/handler.go [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+ */
+
+package adaptor
+
+import (
+	"bytes"
+	"context"
+	"net/http"
+
+	"github.com/cloudwego/hertz/pkg/protocol/consts"
```
Comment (li-jin-gou) [inline pkg/common/adaptor/handler.go:24]: import style

**Pull request closed without merging**